Add no-op benchmarks to test overhead

robieta opened the pull request:
This stack is focused on overhead, but it's not straightforward to actually measure our current overhead. So the first step is to add noop kernels that exercise the machinery, and where effectively 100% of the work can be characterized as "overhead".

Stack from [ghstack](https://github.com/ezyang/ghstack):
* #59393
* #59392
* #59391
* **#59389**

Commit:
Add no-op benchmarks to test overhead

[ghstack-poisoned]

albanD commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -9905,3 +9905,24 @@
   variants: function
   python_module: nn
+
+# Functions which do no work and simply return `self`. Used to measure overhead.
```
If a function return self, then the output alias the input and it should have `Tensor(a)` between them. These signatures are not valid right now.

robieta replied:
This was deliberate, as I wanted to simulate the behavior of a kernel which `could` return a new Tensor (from the point of view of the system); it just happens to do nothing. It might be moot if the perf is basically the same. I'll test that today.

robieta replied:
Turns out it makes no difference, so I've updated the schema. Thanks!

albanD commented on torch/csrc/autograd/VariableTypeManual.cpp:
```diff
@@ -235,6 +235,57 @@ Tensor & detach_(c10::DispatchKeySet ks, Tensor & self) {
   return self;
 }
 
+at::Tensor _noop_unary_manual(c10::DispatchKeySet ks, const at::Tensor & self) {
```
What's the point of having these? They look like they're going to become stale very quickly as it is easy not to update them.

ezyang replied:
They're probably manual because of the soundness problem in https://github.com/pytorch/pytorch/pull/59389/files#r645081077 :)

robieta replied:
I included manual versions because they provide a good way to both see what the kernel is doing and hack on it without having to dive into the codegen. (And rebuilding one cpp file is much faster than the ~700 that get recompiled when you touch codegen.) Staleness is a concern, but I think it should be manageable with proper CI. But I'm open to removing them if folks prefer that.

ezyang replied:
Staleness is a double whammy because if things go stale (but not stale in a functional correctness way), you'll be benchmarking the wrong thing, *silently*

ngimel commented:
Ok, it stems from https://github.com/pytorch/pytorch/pull/58747/files#diff-54b494a4dd0af2160d716378bd5a40e1e4a98c94414901d85adbb6a9ae6dbed2R553-R554, where TI is loading value from the second "input", but second "input" is really output and its memory is uninitialized, so anything can be there. It's no big deal because it's always valid memory, and the value itself is never used, but it throws off asan (because unused value is invalid).

ezyang reviewed (changes requested):
A more reasonable no op operation is to allocate a new tensor (with new storage) and return it. That's the minimum you have to do to properly obey the invariants required by the system.

Commit:
Update on "Add no-op benchmarks to test overhead"

This stack is focused on overhead, but it's not straightforward to actually measure our current overhead. So the first step is to add noop kernels that exercise the machinery, and where effectively 100% of the work can be characterized as "overhead".




[ghstack-poisoned]

robieta commented:
> A more reasonable no op operation is to allocate a new tensor (with new storage) and return it. That's the minimum you have to do to properly obey the invariants required by the system.

Let me make the case for why we don't want to do that. When I started looking into this the baseline instruction count for the unary case was 426 instructions; it is now 504 (an 18% increase), and this is even with our diligent monitoring. Right now we have decent coverage of tensor creation via a slew of `empty` ubenchmarks. This benchmark is intended to be very narrowly focused on the kernel dispatch path and I worry that diluting it with other aspects of the system will make it easier for papercuts to slip in. WDYT?

ezyang commented:
The point of deterministic instruction counts so that dilution doesn't matter: we can still always identify changes no matter how small.

robieta commented:
> The point of deterministic instruction counts so that dilution doesn't matter: we can still always identify changes no matter how small.

Alright, as long as we're confident we'll be disciplined and not fall into "eh, what's a few instructions between friends?". I'll switch it to do a very trivial Tensor creation and update accordingly.

Commit:
Update on "Add no-op benchmarks to test overhead"

This stack is focused on overhead, but it's not straightforward to actually measure our current overhead. So the first step is to add noop kernels that exercise the machinery, and where effectively 100% of the work can be characterized as "overhead".




[ghstack-poisoned]

Commit:
Update on "Add no-op benchmarks to test overhead"

This stack is focused on overhead, but it's not straightforward to actually measure our current overhead. So the first step is to add noop kernels that exercise the machinery, and where effectively 100% of the work can be characterized as "overhead".




[ghstack-poisoned]

Commit:
Update on "Add no-op benchmarks to test overhead"

This stack is focused on overhead, but it's not straightforward to actually measure our current overhead. So the first step is to add noop kernels that exercise the machinery, and where effectively 100% of the work can be characterized as "overhead".




[ghstack-poisoned]

ezyang commented on aten/src/ATen/Overhead.cpp:
```diff
@@ -0,0 +1,37 @@
+namespace native {
+
+Tensor _noop_unary(const Tensor& self) {
+  return std::move(trivial_empty());
```
The `std::move` here is unnecessary for multiple reasons:
1. The expression is already a temporary, so std::move does nothing
2. Even if it did something, RVO is a thing, so you still don't need the move

robieta replied:
Done.

Commit:
Update on "Add no-op benchmarks to test overhead"

This stack is focused on overhead, but it's not straightforward to actually measure our current overhead. So the first step is to add noop kernels that exercise the machinery, and where effectively 100% of the work can be characterized as "overhead".




[ghstack-poisoned]

Commit:
Update on "Add no-op benchmarks to test overhead"

This stack is focused on overhead, but it's not straightforward to actually measure our current overhead. So the first step is to add noop kernels that exercise the machinery, and where effectively 100% of the work can be characterized as "overhead".




[ghstack-poisoned]

Commit:
Update on "Add no-op benchmarks to test overhead"

This stack is focused on overhead, but it's not straightforward to actually measure our current overhead. So the first step is to add noop kernels that exercise the machinery, and where effectively 100% of the work can be characterized as "overhead".




[ghstack-poisoned]

Commit:
Update on "Add no-op benchmarks to test overhead"

This stack is focused on overhead, but it's not straightforward to actually measure our current overhead. So the first step is to add noop kernels that exercise the machinery, and where effectively 100% of the work can be characterized as "overhead".




[ghstack-poisoned]

ezyang commented on aten/src/ATen/native/Overhead.cpp:
```diff
@@ -0,0 +1,29 @@
+
+Tensor _noop_binary(const Tensor& self, const Tensor& other) {
+  return trivial_empty();
+}
```
FWIW, in a previous version you had more overloads, and having a ton of tensor inputs (or a tensor list input) is a pretty valid no-op benchmark, since there are some things we do that have overhead that scale with the number of tensors involved.

robieta replied:
The earlier overloads were for manual kernels, which I removed based on PR feedback. I do like the idea of a tensor list case.

Commit:
Update on "Add no-op benchmarks to test overhead"

This stack is focused on overhead, but it's not straightforward to actually measure our current overhead. So the first step is to add noop kernels that exercise the machinery, and where effectively 100% of the work can be characterized as "overhead".




[ghstack-poisoned]

Commit:
Update on "Add no-op benchmarks to test overhead"

This stack is focused on overhead, but it's not straightforward to actually measure our current overhead. So the first step is to add noop kernels that exercise the machinery, and where effectively 100% of the work can be characterized as "overhead".




[ghstack-poisoned]

Commit:
Update on "Add no-op benchmarks to test overhead"

This stack is focused on overhead, but it's not straightforward to actually measure our current overhead. So the first step is to add noop kernels that exercise the machinery, and where effectively 100% of the work can be characterized as "overhead".




[ghstack-poisoned]

Commit:
Update on "Add no-op benchmarks to test overhead"

This stack is focused on overhead, but it's not straightforward to actually measure our current overhead. So the first step is to add noop kernels that exercise the machinery, and where effectively 100% of the work can be characterized as "overhead".




[ghstack-poisoned]
